### ISSUE TITLE
Signature calculation doesn't consider middleware basePath

### ISSUE THREAD

**Issue by jpike88** [jamhall/s3rver#464]
```
const s3rverInstance = (new S3rver({
		silent: !core.config.verboseLogs,
		directory: awsDir,
		configureBuckets: [{
			name: 'upload/' + aws.filesBucket,
			configs: [`
			<CORSConfiguration>
				<CORSRule>
					<AllowedOrigin>*</AllowedOrigin>
					<AllowedMethod>GET</AllowedMethod>
					<AllowedMethod>POST</AllowedMethod>
					<MaxAgeSeconds>3000</MaxAgeSeconds>
					<AllowedHeader>Authorization</AllowedHeader>
				</CORSRule>
			</CORSConfiguration>
			`]
		}]
	}) as any);

	(async () => {
		await s3rverInstance.run();
		await s3rverInstance.close();
	})();
```

A presigned url generated for the above example looks like this:

```
http://localhost:8081/upload/test-bucket/1%3A4WGxxZdxNC2?AWSAccessKeyId=S3RVER&Content-Disposition=attachment%3B%20filename%3Dlockscreen.png&Content-Type=image%2Fpng&Expires=1558590155&Signature=KylS2DEsoZ3ADb6UQAhs4m9okDo%3D
```

This SHOULD work, but it doesn't. The reason is that the path being used to calculate the string is stripping the prefix of the originalPath.

**Comment by jpike88** [comment]
Current workaround:
```
	const s3rverMiddleware = s3rverInstance.callback();
	app.use('/upload', (req, res, next) => {
		req.url = req.originalUrl;
		s3rverMiddleware(req, res, next);
	});
```

**Comment by jpike88** [comment]
@kherock no easy workaround to this, the above failed when it was on staging with a different domain.. I've downgraded to 2.2.6, as it worked properly before.

**Comment by jpike88** [comment]
@kherock any ideas/comments? This looks like a fairly serious bug...

**Comment by kherock** [comment]
Can you try installing `jamhall/s3rver#fix-baseurl-v2-signatures`? I think it's fixed, I just wanted to verify. I added some tests for using the middleware with Express.

**Comment by jpike88** [comment]
@kherock that branch only works I'm using my server on localhost.

Localhost derived URL (works)...
```http://localhost:8081/upload/test/1%3ABlbzROmB5tL?AWSAccessKeyId=S3RVER&Content-Disposition=attachment%3B%20filename%3D0x0ss%20%281%29.jpg&Content-Type=image%2Fjpeg&Expires=1559640725&Signature=tMQ0z0VPvOGfaLNGFzo9F4qw4Js%3D```

When I deploy to staging however, I get a 403 error.

Staging URL (
```https://staging.api.plantguard.tech/upload/test/14%3AMJMmx6VXYuv?AWSAccessKeyId=S3RVER&Content-Disposition=attachment%3B%20filename%3D0x0ss%20%281%29.jpg&Content-Type=image%2Fjpeg&Expires=1559639467&Signature=Or2HF2k7cAZEVoueFdB1wYMoYuY%3D```

I know the link is https, but it's not using a self-signed certificate.

```
	 const app = express();

	const s3rverInstance = (new S3rver({
		silent: true,
		directory: '/tmp/s3rver_test_directory',
		configureBuckets: [{
			name: 'test,
			configs: [`
			<CORSConfiguration>
				<CORSRule>
					<AllowedOrigin>*</AllowedOrigin>
					<AllowedMethod>GET</AllowedMethod>
					<AllowedMethod>POST</AllowedMethod>
					<MaxAgeSeconds>3000</MaxAgeSeconds>
					<AllowedHeader>Authorization</AllowedHeader>
				</CORSRule>
			</CORSConfiguration>
			`]
		}]
	}) as any);

	app.use('/upload', s3rverInstance.callback());
```

I'm almost certain it's because the signature calculation

**Comment by jpike88** [comment]
I've SSHed into staging and now I understand the issue.

```https://staging.api.plantguard.tech/upload/test/14%3AMJMmx6VXYuv?AWSAccessKeyId=S3RVER&Content-Disposition=attachment%3B%20filename%3D0x0ss%20%281%29.jpg&Content-Type=image%2Fjpeg&Expires=1559639467&Signature=Or2HF2k7cAZEVoueFdB1wYMoYuY%3D```

^ this URL is being incorrectly processed.

I need to use the /upload path for file uploads, because https://staging.api.mysite.tech is my API endpoint.

![Screen Shot 2019-06-04 at 5 37 51 pm](https://user-images.githubusercontent.com/9585787/58873026-8c42bb00-86ef-11e9-816f-7867ae7fac35.png)

ctx.path is initially set as ```/test/2%3ApQ2AKXgjrue```, which is wrong as it's missing the upload path.
ctx.path is then transformed in vhost.js to ```/staging.api.plantguard.tech/test/2%3A3QJLonW2Bin```
Then in authenication.js, canonicalRequest.uri becomes ```/upload/staging.api.plantguard.tech/test/2%3A3QJLonW2Bin```

The domain name ended up past the upload path, which was in the original url.

Further up the stack, in node_modules/koa/lib/application.js:
![Screen Shot 2019-06-04 at 5 44 58 pm](https://user-images.githubusercontent.com/9585787/58873429-8ef1e000-86f0-11e9-8214-240c718ebad0.png)

req.originalURL is correct.

Basically, you're relying on ctx.path being correct, but for calculating the signature, you need more information. You can't trust the ctx variable, you need to crunch the originalURL yourself.

req.originalUrl.split('?')[0] is going to give you a more accurate representation of the original url than ctx.path

**Comment by jpike88** [comment]
Finally...

The signature provided using AWS.S3 was this:

```8ZBFWCCu3HOz8ZuIqsivgSqkgS0=```

I could only match that signature by providing the following string:

"PUT

image/jpeg
1559648124
/upload/test/2%3AKjrMn1M3BFL"
^ This is correct.

"PUT

image/jpeg
1559648124
/upload/staging.api.plantguard.tech/test/2%3AKjrMn1M3BFL"
```
^ This is incorrect.

So that means that vhosts.js is placing the hostname into the path when it's not considered by AWS.S3 originally.

**Comment by jpike88** [comment]
I submitted a PR.

Long story short, virtual host named shouldn't be taken into consideration as unless you provide more context to s3rver, it won't know the different between a virtual host and a normal domain name.

I use s3rver for staging, demo servers etc, and I've confirmed on my PR branch that these fixes are working properly like before. People using weird tricks with virtual host names should consider a different strategy, or submit a PR that allows for an extra bit of configuration. It definitely shouldn't be the default, because it's not the AWS.S3 default.

**Comment by kherock** [comment]
That particular change is actually purely a bugfix, before it would just insert `undefined` into the path, e.g. `/undefined/myKey`. The reason this worked in 2.x should have been because there was no signature checking performed in that version.

Regarding the `originalUrl` stuff, I think adding some early middleware that ensures interoperability with Express's subapp mounting should simplify things.

I'll add some tests around signature calculations with virtual host style URLs and let you know what I run into there. In the meantime, I can see some value in providing a flag to disable signature checking altogether.

P.S. this is a very unique use case that is hitting some edge cases I never considered, so I'm glad all this is being brought to my attention.

**Comment by kherock** [comment]
@jpike88 I think these latest changes in #468 should be it - I've released the workaround as-is in v3.2.0! If you still can't pass signature checks, I've added the `allowMismatchedSignatures` option to S3rver's constructor.

**Comment by jpike88** [comment]
nice, I'll give it a run on staging over the next day and let you know the results
